File: commons-parsers/src/main/java/xyz/wagyourtail/commons/parsers/impl/json5/Json5Value.java
package xyz.wagyourtail.commons.parsers.impl.json5;

import lombok.val;
import xyz.wagyourtail.commons.core.reader.CharReader;
import xyz.wagyourtail.commons.core.reader.StringCharReader;
import xyz.wagyourtail.commons.parsers.Data;
import xyz.wagyourtail.commons.parsers.StringData;
import xyz.wagyourtail.commons.parsers.impl.constant.BooleanConstant;

public class Json5Value extends StringData.OnlyParsed<Data.SingleContent<?>> {

    protected Json5Value(SingleContent<?> content) {
        super(content);
    }

    public static Json5Value parse(String raw) {
        StringCharReader charReader = new StringCharReader(raw);
        val content = contentBuilder(charReader);
        return new Json5Value(content);
    }

    public static Json5Value parse(CharReader<?> reader) {
        return new Json5Value(contentBuilder(reader));
    }

    public static SingleContent<?> contentBuilder(CharReader<?> reader) {
        return new SingleContent<>(reader.<Object>parse(
                "json5 value",
                Json5Object::parse,
                Json5Array::parse,
                Json5StringConstant::new,
                BooleanConstant::new,
                Json5Number::new,
                r -> {
                    r.expect("null");
                    return null;
                }
        ));
    }

    public Object getValue() {
        return getContent().getValue();
    }

}

File: commons-parsers-kt-kt/src/commonMain/kotlin/xyz/wagyourtail/commonskt/parsers/impl/json5/Json5Value.kt
package xyz.wagyourtail.commonskt.parsers.impl.json5

import xyz.wagyourtail.commonskt.parsers.Data
import xyz.wagyourtail.commonskt.parsers.StringData
import xyz.wagyourtail.commonskt.parsers.impl.constant.BooleanConstant
import xyz.wagyourtail.commonskt.reader.CharReader

class Json5Value(content: SingleContent<*>) : StringData.OnlyParsed<Data.SingleContent<*>>(content) {

    companion object : StringDataBuilder<Json5Value, Data.SingleContent<*>> {

        override fun invoke(rawContent: CharReader<*>): Json5Value {
            return Json5Value(checkedBuildContent(rawContent))
        }

        override fun checkedBuildContent(reader: CharReader<*>): SingleContent<*> {
            return SingleContent<Any?>(reader.parse(
                "json5 value",
                Json5Object::invoke,
                Json5Array::invoke,
                Json5StringConstant::invoke,
                BooleanConstant::invoke,
                Json5Number::invoke,
                {
                    expect("null")
                    null
                }
            ))
        }

    }

}